pandas/core/construction.py

```diff
@@ -270,7 +275,7 @@ def array(
         return cls._from_sequence(data, dtype=dtype, copy=copy)
 
     if dtype is None:
-        inferred_dtype = lib.infer_dtype(data, skipna=False)
+        inferred_dtype = lib.infer_dtype(data, skipna=True)
```

Comment:
so my issue with this PR is that is duplicating a lot of logic that is already held here: https://github.com/pandas-dev/pandas/blob/master/pandas/_libs/lib.pyx#L1948, so now we have 2 places with slightly different ways of doing things.

This routine is slightly more 'high-level', but myabe_convert_objects is way more used internally. So how to reconcile these things?

@jorisvandenbossche @TomAugspurger

Reply:
I'm not familiar with `maybe_convert_objects`, but what's the duplicate logic between the two? At a glance, it seems like `maybe_convert_objects` is mixing two things

1. type inference (the sole purpose of `lib.infer_dtype`)
2. array construction (which may be `pd.array`? Or move the core to some `_libs` method?).

Should we update `maybe_infer_objects` to use `array` internally? I don't have a feel for whether that's even possible, do you?

Reply:
> Should we update maybe_infer_objects to use array internally? I don't have a feel for whether that's even possible, do you?

Probably not in the short term.  Best guess for de-duplication for the discussed functions will look something like:

- `lib.maybe_convert_objects` is made to back `lib.infer_dtype`
  - Note: `lib.maybe_convert_objects` involves two runtime non-cython imports that I'd really like to find a way to avoid (one will be easy, the other very much not)
- `pd.array` calls `maybe_convert_objects` instead of `infer_dtype`
  - More generally, many places where we call infer_dtype followed by casting can be replaced to be one-pass instead of two-pass.
- `Series` constructor is backed by `pd.array`
- `Index.__new__` is backed by `pd.array` (we'd need something like `RangeArray` first)